### PR TITLE
carousel upgrade notice

### PR DESCRIPTION
**Pull request by michaelkirk**
Notify the user of recent changes.

**iPhone 6 - iOS 10.1**
iphone 6 - screen 1

iphone 6 - screen 2

**iPhone 5 - iOS 9.3** (note, no callkit screen for iOS 9)
iphone5 - screen 1

**iPhone 7+ - iOS 10.2**
iphone7 - screen 1

iphone7 - screen 2

** Privacy screen changes **

Group the two callkit settings together to better communicate that they are dependent.

screen shot 2017-03-02 at 3 02 30 pm

PTAL @charlesmchen

### CODE REVIEW

**Comment by charlesmchen** [comment]
* Dismiss button looks small on 7+.  
* I hope the hot area for dismiss is much larger than the "dismiss" text.
* I like the combined privacy settings.

**Comment by charlesmchen** [comment]
"Read more in your privacy settings".  We could make this text hot, opening the privacy settings, and indicate that by underlining+"signal blue" the portion "privacy settings", like a link.

File: Signal/src/view controllers/ExperienceUpgradeViewController.swift [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+    private func addDismissGesture() {
+        let swipeGesture = UISwipeGestureRecognizer(target: self, action: #selector(handleDismissGesture))
+        swipeGesture.direction = .down
+        view.addGestureRecognizer(swipeGesture)
+    }
+
+    // MARK: Carousel
+
+    private var carouselView: UIScrollView!
+    private var currentSlideIndex = 0
```
Comment (charlesmchen) [inline Signal/src/view controllers/ExperienceUpgradeViewController.swift]: Huh, so you didn't want to gather these at the top of the class.

Did you consider using `UIPageViewController`?  It solves a very similar problem pretty well.

**Comment by riyapenn** [comment]
I understand the purpose of the green video icon is to demonstrate a video call, but it does not reflect the UIUX.  The video call icon in green should display as a red handset icon with mute/video next to it (if displayed) (along w/ name/number and call duration).

File: Signal/src/view controllers/ExperienceUpgradeViewController.swift [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+
+        // Next button layout
+        nextButton.autoPinEdge(toSuperviewEdge: .right)
+        nextButton.autoPinEdge(toSuperviewEdge: .top, withInset: arrowButtonInset)
+
+        // Dismiss button
+        let dismissButton = UIButton()
+        splashView.addSubview(dismissButton)
+        dismissButton.setTitle(NSLocalizedString("DISMISS_BUTTON_TEXT", comment: ""), for: .normal)
+        dismissButton.setTitleColor(UIColor.ows_materialBlue(), for: .normal)
```
Comment (charlesmchen) [inline Signal/src/view controllers/ExperienceUpgradeViewController.swift]: When do we use `materialBlue` and when do we use the "signal blue"?

**Comment by michaelkirk** [comment]
> I understand the purpose of the green video icon is to demonstrate a video call, but it does not reflect the UIUX. The video call icon in green should display as a red handset icon with mute/video next to it (if displayed) (along w/ name/number and call duration).

That was deliberate - I wanted to highlight the video button, and adding *all* the interface would be a lot of noise. I don't expect it will confuse anybody.

File: Signal/src/view controllers/ExperienceUpgradeViewController.swift [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+        previousButton.autoPinEdge(toSuperviewEdge: .top, withInset: arrowButtonInset)
+
+        // Next button
+        let nextButton = UIButton()
+        self.nextButton = nextButton
+        splashView.addSubview(nextButton)
+        nextButton.isUserInteractionEnabled = true
+        nextButton.setTitleColor(arrowButtonColor, for: .normal)
+        nextButton.accessibilityLabel = NSLocalizedString("UPGRADE_CAROUSEL_NEXT_BUTTON", comment: "accessibility label for arrow in slideshow")
+        nextButton.setTitle("›", for: .normal)
```
Comment (charlesmchen) [inline Signal/src/view controllers/ExperienceUpgradeViewController.swift]: Just FYI - Another option is to use the iOS < "back/prev" and > "next" png assets.  I've already extracted and added the former as "NavBarBack".  NBD, but I think they'd read as more idiomatic.

I used some fork of https://github.com/futuretap/iOS-Artwork-Extractor to do the extraction.

File: Signal/src/view controllers/ExperienceUpgradeViewController.swift [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+        let containerPadding = ScaleFromIPhone5To7Plus(12, 24)
+        slideView.autoPinEdgesToSuperviewEdges(with: UIEdgeInsets(top: containerPadding, left: containerPadding, bottom: containerPadding, right: containerPadding))
+
+        // Body label
+        let bodyLabel = UILabel()
+        slideView.addSubview(bodyLabel)
+        bodyLabel.text = body
+        bodyLabel.font = UIFont.ows_lightFont(withSize: ScaleFromIPhone5To7Plus(18, 22))
+        bodyLabel.textColor = UIColor.black
+        bodyLabel.numberOfLines = 0
```
Comment (charlesmchen) [inline Signal/src/view controllers/ExperienceUpgradeViewController.swift]: Whenever you set `numberOfLines` to zero, you probably want: `bodyLabel.lineBreakMode = .byWordWrapping`.

File: Signal/src/environment/ExperienceUpgradeFinder.swift [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+//
+
+import Foundation
+
+class ExperienceUpgradeFinder: NSObject {
+    public let TAG = "[ExperienceUpgradeFinder]"
+
+    // Keep these ordered by increasing uniqueId.
+    private var allExperienceUpgrades: [ExperienceUpgrade] {
+        var upgrades = [ExperienceUpgrade(uniqueId: "001",
```
Comment (michaelkirk) [inline Signal/src/environment/ExperienceUpgradeFinder.swift]: done

File: Signal/src/environment/ExperienceUpgradeFinder.swift [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+            upgrades.append(ExperienceUpgrade(uniqueId: "002",
+                                              title: NSLocalizedString("UPGRADE_EXPERIENCE_CALLKIT_TITLE", comment: "Header for upgrade experience"),
+                                              body: NSLocalizedString("UPGRADE_EXPERIENCE_CALLKIT_DESCRIPTION", comment: "Description of CallKit to upgrading (existing) users"),
+                                              image: #imageLiteral(resourceName: "introductory_splash_callkit")))
+        }
+
+        return upgrades
+    }
+
+
```
Comment (michaelkirk) [inline Signal/src/environment/ExperienceUpgradeFinder.swift]: done

File: Signal/src/view controllers/ExperienceUpgradeViewController.swift (hunk 4 above)
Comment (michaelkirk) [inline Signal/src/view controllers/ExperienceUpgradeViewController.swift]: done

**Comment by michaelkirk** [comment]
> Dismiss button looks small on 7+.
> I hope the hot area for dismiss is much larger than the "dismiss" text.

Fixed.

> When do we use materialBlue and when do we use the "signal blue"?

Good question - materialBlue is used for all the big blocks of blue color. IIRC signal blue was extracted by you from some hardcoded value in the storyboard where it was used to color labels. SignalBlue is roughly the same mix of colors as MaterialBlue, but a little darker. 

I propose that the (darker) signalBlue should be used for thin stroke text, whereas the (slightly lighter) materialBlue should be used for coloring larger blocks (like the navbar)

I'll update the dismiss label accordingly, to use the darker of the pair, and rename the colors to `signalBlueBackground` and `signalBlueText` to avoid future confusion.

**Comment by michaelkirk** [comment]
Some substantial changes in

https://github.com/WhisperSystems/Signal-iOS/pull/1823/commits/9b2eb80392b6a50af0923263e1b082135609562b
PTAL @charlesmchen

**Comment by NicolasCGN** [comment]
Would it make sense now to move the menu item regarding the message notification style (show name and message) from 'Advanced' into the 'Privacy' section?

**Pull request merged**